# Phase 6: Model Optimization

## Objective

The objective of this phase is to improve the predictive performance of the selected machine learning model through hyperparameter tuning.

Based on the evaluation results obtained in the previous phase, XGBoost was identified as the best-performing model. Therefore, it is selected for optimization using RandomizedSearchCV to identify the optimal combination of hyperparameters.

The optimized model will then be evaluated and compared with the original model to determine whether its performance has improved.

---

## Tasks

- Load the selected XGBoost model.
- Perform hyperparameter tuning using RandomizedSearchCV.
- Train the optimized XGBoost model.
- Evaluate the optimized model.
- Compare the optimized model with the original model.

In [1]:
# ==========================================================
# Step 1: Import Required Libraries
# ==========================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

# XGBoost
from xgboost import XGBClassifier

# Hyperparameter Tuning
from sklearn.model_selection import RandomizedSearchCV

# Evaluation Metrics
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    ConfusionMatrixDisplay
)

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
# ==========================================================
# Step 2: Load Processed Datasets
# ==========================================================

# Load training data
X_train = pd.read_csv("../data/processed/X_train.csv")
y_train = pd.read_csv("../data/processed/y_train.csv").squeeze()

# Load testing data
X_test = pd.read_csv("../data/processed/X_test.csv")
y_test = pd.read_csv("../data/processed/y_test.csv").squeeze()

print("Processed datasets loaded successfully.\n")

print("Training Features :", X_train.shape)
print("Training Target   :", y_train.shape)

print("Testing Features  :", X_test.shape)
print("Testing Target    :", y_test.shape)

Processed datasets loaded successfully.

Training Features : (54564, 12)
Training Target   : (54564,)
Testing Features  : (13641, 12)
Testing Target    : (13641,)


## Observation

The processed training and testing datasets were successfully loaded for model optimization.

These datasets will be used to perform hyperparameter tuning on the selected XGBoost model using RandomizedSearchCV. The training dataset will be used to identify the optimal hyperparameter values, while the testing dataset will be used to evaluate the performance of the optimized model.

In [3]:
# ==========================================================
# Step 3: Define Hyperparameter Search Space
# ==========================================================

param_grid = {

    # Number of boosting trees
    "n_estimators": [100, 200, 300, 500],

    # Maximum depth of each tree
    "max_depth": [3, 5, 7, 9],

    # Learning rate
    "learning_rate": [0.01, 0.05, 0.1, 0.2],

    # Fraction of training samples used for each tree
    "subsample": [0.7, 0.8, 0.9, 1.0],

    # Fraction of features used for each tree
    "colsample_bytree": [0.7, 0.8, 0.9, 1.0],

    # Minimum loss reduction required for a split
    "gamma": [0, 0.1, 0.3, 0.5],

    # Minimum sum of instance weight required in a child
    "min_child_weight": [1, 3, 5]
}

print("Hyperparameter search space defined successfully.")

Hyperparameter search space defined successfully.


## Observation

A hyperparameter search space was defined for the XGBoost classifier. The selected hyperparameters influence the complexity, learning behavior, and generalization capability of the model.

The search space includes parameters such as the number of trees, tree depth, learning rate, sampling ratios, and split criteria. These values will be explored during hyperparameter tuning using RandomizedSearchCV to identify the optimal combination for improved prediction performance.

In [4]:
# ==========================================================
# Step 4: Perform Hyperparameter Tuning
# ==========================================================

random_search = RandomizedSearchCV(

    estimator=XGBClassifier(
        random_state=42,
        eval_metric="logloss"
    ),

    param_distributions=param_grid,

    n_iter=20,

    scoring="accuracy",

    cv=5,

    verbose=2,

    random_state=42,

    n_jobs=-1
)

random_search.fit(X_train, y_train)

print("Hyperparameter tuning completed successfully.")

Fitting 5 folds for each of 20 candidates, totalling 100 fits
[CV] END colsample_bytree=0.9, gamma=0.1, learning_rate=0.05, max_depth=9, min_child_weight=3, n_estimators=200, subsample=0.9; total time=   0.9s
[CV] END colsample_bytree=0.9, gamma=0.1, learning_rate=0.05, max_depth=9, min_child_weight=3, n_estimators=200, subsample=0.9; total time=   0.9s
[CV] END colsample_bytree=0.9, gamma=0.1, learning_rate=0.05, max_depth=9, min_child_weight=3, n_estimators=200, subsample=0.9; total time=   0.9s
[CV] END colsample_bytree=0.9, gamma=0.1, learning_rate=0.05, max_depth=9, min_child_weight=3, n_estimators=200, subsample=0.9; total time=   0.9s
[CV] END colsample_bytree=0.9, gamma=0.1, learning_rate=0.05, max_depth=9, min_child_weight=3, n_estimators=200, subsample=0.9; total time=   0.9s
[CV] END colsample_bytree=0.7, gamma=0.1, learning_rate=0.01, max_depth=5, min_child_weight=5, n_estimators=500, subsample=0.7; total time=   1.3s
[CV] END colsample_bytree=0.7, gamma=0.1, learning_rate=

## Observation

Hyperparameter tuning was performed using **RandomizedSearchCV** with 5-fold cross-validation.

Twenty different combinations of XGBoost hyperparameters were evaluated to identify the configuration that provides the highest classification accuracy.

This optimization process helps improve the model's generalization performance while reducing the risk of overfitting.

In [5]:
# ==========================================================
# Step 5: Retrieve the Best Hyperparameters
# ==========================================================

# Best hyperparameter combination
print("Best Hyperparameters:\n")
print(random_search.best_params_)

print("\n" + "="*60)

# Best cross-validation accuracy
print(f"Best Cross-Validation Accuracy: {random_search.best_score_:.4f}")

print("\n" + "="*60)

# Optimized XGBoost model
optimized_xgb = random_search.best_estimator_

print("Optimized XGBoost model created successfully.")

Best Hyperparameters:

{'subsample': 0.8, 'n_estimators': 300, 'min_child_weight': 3, 'max_depth': 7, 'learning_rate': 0.01, 'gamma': 0, 'colsample_bytree': 0.9}

Best Cross-Validation Accuracy: 0.7341

Optimized XGBoost model created successfully.


## Observation

RandomizedSearchCV successfully identified the optimal hyperparameter combination for the XGBoost classifier.

The best hyperparameters were selected based on 5-fold cross-validation, resulting in a cross-validation accuracy of **73.41%**.

The optimized XGBoost model was created using these hyperparameters and will now be evaluated on the testing dataset to determine whether the optimization has improved its predictive performance compared to the original model.

In [6]:
# ==========================================================
# Step 6: Evaluate the Optimized XGBoost Model
# ==========================================================

# Generate predictions
y_pred_optimized = optimized_xgb.predict(X_test)

# Calculate evaluation metrics
optimized_accuracy = accuracy_score(y_test, y_pred_optimized)
optimized_precision = precision_score(y_test, y_pred_optimized)
optimized_recall = recall_score(y_test, y_pred_optimized)
optimized_f1 = f1_score(y_test, y_pred_optimized)
optimized_roc_auc = roc_auc_score(y_test, y_pred_optimized)

print("=" * 60)
print("Optimized XGBoost Model Performance")
print("=" * 60)

print(f"Accuracy   : {optimized_accuracy:.4f}")
print(f"Precision  : {optimized_precision:.4f}")
print(f"Recall     : {optimized_recall:.4f}")
print(f"F1-Score   : {optimized_f1:.4f}")
print(f"ROC-AUC    : {optimized_roc_auc:.4f}")

Optimized XGBoost Model Performance
Accuracy   : 0.7377
Precision  : 0.7609
Recall     : 0.6834
F1-Score   : 0.7201
ROC-AUC    : 0.7370


## Observation

The optimized XGBoost model was evaluated on the testing dataset using the same performance metrics as the original model.

The optimized model achieved an Accuracy of **73.77%**, Precision of **76.09%**, Recall of **68.34%**, F1-Score of **72.01%**, and ROC-AUC Score of **73.70%**.

Compared to the original XGBoost model, the optimized model showed slight improvements in Accuracy, Precision, F1-Score, and ROC-AUC Score, while the Recall decreased marginally. Overall, hyperparameter tuning resulted in a modest improvement in the model's predictive performance.

In [7]:
# ==========================================================
# Step 7: Compare Original and Optimized XGBoost Models
# ==========================================================

comparison = pd.DataFrame({

    "Metric": [
        "Accuracy",
        "Precision",
        "Recall",
        "F1-Score",
        "ROC-AUC"
    ],

    "Original XGBoost": [
        0.7362,
        0.7581,
        0.6838,
        0.7191,
        0.7356
    ],

    "Optimized XGBoost": [
        optimized_accuracy,
        optimized_precision,
        optimized_recall,
        optimized_f1,
        optimized_roc_auc
    ]
})

comparison["Improvement"] = (
    comparison["Optimized XGBoost"] -
    comparison["Original XGBoost"]
)

comparison

,Metric,Original XGBoost,Optimized XGBoost,Improvement
0,Accuracy,0.7362,0.737703,0.001503
1,Precision,0.7581,0.760913,0.002813
2,Recall,0.6838,0.683398,-0.000402
3,F1-Score,0.7191,0.720075,0.000975
4,ROC-AUC,0.7356,0.737022,0.001422


## Observation

A comparison was performed between the original and optimized XGBoost models using five evaluation metrics.

The optimized model demonstrated improvements in Accuracy, Precision, F1-Score, and ROC-AUC Score, indicating that hyperparameter tuning enhanced the model's overall predictive performance. Although the Recall decreased slightly, the change was minimal and did not significantly affect the model's effectiveness.

Based on the comparison, the optimized XGBoost model is selected as the final model for heart disease prediction due to its superior overall performance.

In [8]:
# ==========================================================
# Step 8: Save the Optimized XGBoost Model
# ==========================================================

import joblib

joblib.dump(
    optimized_xgb,
    "../models/optimized_xgboost.pkl"
)

print("Optimized XGBoost model saved successfully.")

Optimized XGBoost model saved successfully.


## Observation

The optimized XGBoost model was successfully saved in the **models** directory using Joblib.

Saving the trained model allows it to be reused later for prediction without retraining, making deployment and future inference more efficient.

# 📌 Phase 6 Summary

## Objective Achieved

The objective of this phase was to improve the performance of the best-performing machine learning model using hyperparameter optimization.

### Tasks Completed

- Selected the best-performing model from Phase 5 (XGBoost).
- Defined a hyperparameter search space for optimization.
- Performed hyperparameter tuning using RandomizedSearchCV.
- Identified the optimal hyperparameter combination.
- Retrained the XGBoost model using the best parameters.
- Evaluated the optimized model on the test dataset.
- Compared the optimized model with the original XGBoost model.
- Saved the optimized model for future use.

### Best Hyperparameters

- n_estimators = **300**
- learning_rate = **0.01**
- max_depth = **7**
- min_child_weight = **3**
- subsample = **0.8**
- colsample_bytree = **0.9**
- gamma = **0**

### Final Performance

| Metric | Value |
|---------|------:|
| Accuracy | **73.77%** |
| Precision | **76.09%** |
| Recall | **68.34%** |
| F1-Score | **72.01%** |
| ROC-AUC | **73.70%** |

### Conclusion

Hyperparameter tuning slightly improved the performance of the XGBoost model. The optimized XGBoost model achieved the best overall balance across the evaluation metrics and is selected as the final machine learning model for heart disease prediction.